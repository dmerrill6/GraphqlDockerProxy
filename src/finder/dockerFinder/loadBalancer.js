import express from 'express'
import request from 'request'

let lBserver = []

export const loadANewLoadBalaceMiddleware = (listOfBackends) => {

  const servers = listOfBackends.map(one => {
    return one.url
  })
  let cur = 0

  const handler = (req, res) => {

    // console.log('hier:', req.headers, servers.length)
    if (req.url === '/'){
      req.url = ''
    }
    req.pipe(request({ url: servers[cur] + req.url })).pipe(res)
    cur = (cur + 1) % servers.length
  }
  const server = express()
  server.get('*', handler).post('*', handler)
  server.post('*', handler).post('*', handler)
  const port = getNextPortNumber()
  lBserver.push(server.listen(port))
  return {
    url: 'http://127.0.0.1:' + port + '',
    clients: listOfBackends.map((one) => Object.assign({}, one)),
  }
}

let nextPortNumber = 0
const getNextPortNumber = () => {
  nextPortNumber = nextPortNumber + 1
  return 8000 + nextPortNumber
}

export const closeAllServer = () => {
  lBserver.forEach(one => {
    console.log('close load balancer', one._connectionKey)
    one.close()
  })
  nextPortNumber = 0
  lBserver = []
}
