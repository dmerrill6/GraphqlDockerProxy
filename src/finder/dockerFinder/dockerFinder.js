//@flow
import { Docker } from 'node-docker-api'
import { token, network } from '../../properties'
import { loadANewLoadBalaceMiddleware, closeAllServer } from './loadBalancer'
import { FindEndpoints } from '../findEndpointsInterface'
import type { Endpoints } from '../findEndpointsInterface'
import * as clientLabels from '../clientLabels'
import { sortEndpointAndFindAvailableEndpoints } from '../endpointsAvailable'
export class DockerFinder extends FindEndpoints{
  constructor(){
    super()
  }

  handleRestart = (endpoints : Endpoints): Endpoints => {
    return this.foundEquals(endpoints)
  }


  foundEquals = async(data: Endpoints) :Endpoints => {
    data = await sortEndpointAndFindAvailableEndpoints(data)
    closeAllServer()
    for (const one in data){
      const namespace = data[one]
      for (let i = 0, l = namespace.length; i < l; i++){
        const lbData = {}
        const searchingElement = namespace[i]
        if (searchingElement.__burnd === undefined){
          for (let j = i + 1; j < l; j++){
            const testingElement = namespace[j]

            if (searchingElement.__imageID == testingElement.__imageID){
              if (lbData[searchingElement.__imageID] === undefined){
                lbData[searchingElement.__imageID] = []
              }
              lbData[searchingElement.__imageID].push(testingElement)
              namespace[j].__burnd = true
            // namespace.splice(j)
            }


          }
          if (lbData[searchingElement.__imageID] !== undefined){
            lbData[searchingElement.__imageID].push(searchingElement)
            const lbResult = loadANewLoadBalaceMiddleware(lbData[searchingElement.__imageID])
            namespace[i].url = lbResult.url
            namespace[i].__loadbalance = {
              count: namespace.length,
              endpoints: lbResult.clients,
            }

          }
        }
        const newNamespaces = []
        for (const oneBurable in namespace){
          if (!namespace[oneBurable].__burnd){
            newNamespaces.push(namespace[oneBurable])
          }
        }
        data[one] = newNamespaces

      }

    }

    return data
  }

  /**
 * Schaut ob es sich um eine absolute url handelt.(Startet mit http(s)://)
 * Wenn nicht sucht sie die ip des netzwerkes raus. (Relative URL z.B.: :{port}{suburl}(:3001/graphql))
 * 
 * 
 */
 updateUrl = (url:string, sockData:any) :string => {
   if (url.startsWith('http')){
     return url
   } else {
     return 'http://' + sockData.NetworkSettings.Networks[network()].IPAddress + url
   }
 }

  getEndpoints = async(): Endpoints => {
    const docker = new Docker({ socketPath: '/var/run/docker.sock' })

    return await docker.container.list().then(containers => {
      const result = {}
      containers.forEach(one => {
        if (one.data.Labels[clientLabels.TOKEN] == token()){
          const url = one.data.Labels[clientLabels.URL]
          const namespace = one.data.Labels[clientLabels.NAMESPACE]
          if (result[namespace] == undefined){
            result[namespace] = []
          }
          result[namespace].push({
            url: this.updateUrl(url, one.data),
            namespace,
            typePrefix: namespace + '_',
            __created: one.data.Created,
            __imageID: one.data.Image,
          })
        } else {
          // console.log('no gqlProxy labels set')
        }
      })


      return result
    })
  }
}

