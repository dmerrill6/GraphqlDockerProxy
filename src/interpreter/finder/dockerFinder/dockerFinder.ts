import { Docker } from 'node-docker-api';
import { token, network } from '../../../properties';
import { loadANewLoadBalaceMiddleware, closeAllServer } from './loadBalancer';
import { FindEndpoints } from '../findEndpointsInterface';
import { Endpoints } from '../../endpoints';
import * as clientLabels from '../../clientLabels';
import { sortEndpointAndFindAvailableEndpoints } from '../../endpointsAvailable';
export class DockerFinder extends FindEndpoints{
  constructor() {
    super();
  }

  handleRestart = (endpoints : Endpoints): Promise<Endpoints> => {
    return this.foundEquals(endpoints);
  }

  foundEquals = async(datas: Endpoints) :Promise<Endpoints> => {
    const data = await sortEndpointAndFindAvailableEndpoints(datas);
    closeAllServer();
    for (const one in data) {
      const namespace = data[one];
      for (let i = 0, l = namespace.length; i < l; i = i + 1) {
        const lbData = {};
        const searchingElement = namespace[i];
        if (searchingElement.__burnd === undefined) {
          for (let j = i + 1; j < l; j = j + 1) {
            const testingElement = namespace[j];

            if (searchingElement.__imageID === testingElement.__imageID) {
              if (lbData[searchingElement.__imageID] === undefined) {
                lbData[searchingElement.__imageID] = [];
              }
              lbData[searchingElement.__imageID].push(testingElement);
              namespace[j].__burnd = true;
            // namespace.splice(j)
            }

          }
          if (lbData[searchingElement.__imageID] !== undefined) {
            lbData[searchingElement.__imageID].push(searchingElement);
            const lbResult = loadANewLoadBalaceMiddleware(lbData[searchingElement.__imageID]);
            namespace[i].url = lbResult.url;
            namespace[i].__loadbalance = {
              count: namespace.length,
              endpoints: lbResult.clients,
            };

          }
        }
        const newNamespaces = [];
        for (const oneBurable in namespace) {
          if (!namespace[oneBurable].__burnd) {
            newNamespaces.push(namespace[oneBurable]);
          }
        }
        data[one] = newNamespaces;

      }

    }

    return data;
  }

  /**
 * Schaut ob es sich um eine absolute url handelt.(Startet mit http(s)://)
 * Wenn nicht sucht sie die ip des netzwerkes raus.
 * (Relative URL z.B.: :{port}{suburl}(:3001/graphql))
 *
 *
 */
  updateUrl = (url:string, sockData:any) :string => {
    if (url.startsWith('http')) {
      return url;
    }
    return 'http://' + sockData.NetworkSettings.Networks[network()].IPAddress + url;

  }

  getEndpoints = async(): Promise<Endpoints> => {
    const docker = new Docker({ socketPath: '/var/run/docker.sock' });

    return await docker.container.list().then((containers) => {
      const result = {};
      containers.forEach((one) => {
        if (idx(one, _ => _.data.Labels[clientLabels.TOKEN]) + '' === token()) {
          const url = idx(one, _ => _.data.Labels[clientLabels.URL]);
          const namespace = idx(one, _ => _.data.Labels[clientLabels.NAMESPACE]);
          if (result[namespace] === undefined) {
            result[namespace] = [];
          }

          result[namespace].push({
            namespace,
            url: this.updateUrl(url, one.data),
            typePrefix: namespace + '_',
            __created: idx(one, _ => _.data.Created),
            __imageID: idx(one, _ => _.data.Image),
          });
        } else {
          // console.log('no gqlProxy labels set')
        }
      });

      return result;
    });
  }
}
