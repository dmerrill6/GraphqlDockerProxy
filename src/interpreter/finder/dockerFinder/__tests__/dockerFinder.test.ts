import { DockerFinder } from '../dockerFinder';
import { Endpoints } from '../../../endpoints';
jest.mock('../../../endpointsAvailable', () => {
  return {
    sortEndpointAndFindAvailableEndpoints: (data) => {
      return Promise.resolve(data);
    },
  };
});
jest.mock('../../../loadBalancer', () => {
  return {
    foundEquals: (endpoints: Endpoints): Promise<Endpoints> => {
      return Promise.resolve(endpoints);
    },
    closeAllServer: () => {},
    loadANewLoadBalaceMiddleware: (listOfBackends) => {
      return {
        url: 'http://127.0.0.1:8001',
        clients: listOfBackends.map((one) => {
          return Object.assign({}, one);
        }),
      };

    },
  };
});
jest.mock('../../../../properties', () => {
  return {
    token: () => {
      return '1234';
    },
    network: () => {
      return 'web';
    },
  };
});
jest.mock('node-docker-api');
describe('Tests the docá¸±erfinder Class', () => {
  let dockerFinder = null;
  beforeEach(() => {
    dockerFinder = new DockerFinder();
  });

  describe('tests method getEndpoints', () => {
    xit('gives a list of object and filter the right endpoints', async() => {
      const endpoints = await dockerFinder.getEndpoints();
      expect(endpoints).toMatchSnapshot();
    });
  });

  describe('tests handleRestart', () => {
    xit('tests there does not need loadbalance no change at endpoints ', async() => {
      const endpoints = await dockerFinder.getEndpoints();
      const newEndpoints = await dockerFinder.handleRestart(endpoints);
      expect(endpoints).toEqual(newEndpoints);
    });

    xit('tests there need loadbalance endpoints change ', async() => {
      const endpoints = await dockerFinder.getEndpoints();
      endpoints['one'].push({
        url: 'http://124.122.123.123:8000/graphql',
        namespace: 'one',
        typePrefix: 'one_',
        created: '20180101',
        __imageID: 'one',
      });
      const newEndpoints = await dockerFinder.handleRestart(endpoints);
      expect(newEndpoints['one'][0]).toMatchSnapshot();

    });

  });
  describe('tests updateUrl ', () => {
    it('by absolut url', () => {
      const url = 'https://test.de/graphql';
      expect(dockerFinder.updateUrl(url, {})).toBe(url);
    });
    it('by relativ url', () => {
      const sockData = {
        NetworkSettings: {
          Networks: {
            web: {
              IPAddress: '127.0.0.1',
            },
          },
        },
      };
      expect(dockerFinder.updateUrl(':3000/graphql', sockData))
      .toBe('http://127.0.0.1:3000/graphql');
    });
  });

});
