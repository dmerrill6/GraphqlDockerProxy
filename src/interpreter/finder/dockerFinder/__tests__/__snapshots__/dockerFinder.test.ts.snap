// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests the docḱerfinder Class tests handleRestart tests there need loadbalance endpoints change  1`] = `
Object {
  "__created": "20180101",
  "__imageID": "one",
  "__loadbalance": Object {
    "count": 2,
    "endpoints": Array [
      Object {
        "__burnd": true,
        "__imageID": "one",
        "created": "20180101",
        "namespace": "one",
        "typePrefix": "one_",
        "url": "http://124.122.123.123:8000/graphql",
      },
      Object {
        "__created": "20180101",
        "__imageID": "one",
        "namespace": "one",
        "typePrefix": "one_",
        "url": "http://123.122.123.123:9000/graphql",
      },
    ],
  },
  "namespace": "one",
  "typePrefix": "one_",
  "url": "http://127.0.0.1:8001",
}
`;

exports[`Tests the docḱerfinder Class tests method getEndpoints gives a list of object and filter the right endpoints 1`] = `Object {}`;
